source/common/grpc/google_async_client_impl.h

```diff
@@ -178,10 +205,22 @@ class GoogleAsyncStreamImpl : public AsyncStream,
   bool write_pending_{};
   // Is an Operation::Finish in-flight?
   bool finish_pending_{};
-  // Is a cq_.Shutdown() in-progress?
-  bool cq_shutdown_in_progress_{};
+  // Have we entered CQ draining state? If so, we're just waiting for all our
+  // ops on the CQ to drain away before freeing the stream.
+  // We need to access this cross-thread. The rationale for this is that the
+  // GoogleAsyncStreamImpl silo thread can safely set and check while draining
+  // the status. However, if the completion thread does not check this before
+  // posting, it might post a new completion after cleanup() is invoked. So, we
+  // need to share state here to allow the completion thread to avoid doing
+  // this.
+  bool draining_cq_{}; // Guarded by cq_lock.
```

Comment:
Can we use the clang thread annotations potentially?

Reply:
I'd like to do this. Do you know if the Clang Mac OS X issues that we previously hit are still a thing?

Reply:
I think the only reason I had an issue was due to using C++14 R/W lock . I think you should probably be fine.